File: 250321/마을 구분하기/seperate-village.cpp
#include <bits/stdc++.h>
using namespace std;

#define MAX_N 25
#define DIR_NUM 4

int n;
int grid[MAX_N][MAX_N];
bool visited[MAX_N][MAX_N];

vector<int> people;
int cnt = 0;

bool CanGo(int x, int y) {
    return 0<=x && x<n && 0<=y && y<n && grid[x][y] && !visited[x][y];
}

void DFS(int x, int y) {
    int dx[DIR_NUM] = {-1, 0, 1, 0}, dy[DIR_NUM] = {0, 1, 0, -1};

    for(int d=0; d<DIR_NUM; d++) {
        int nx = x+dx[d], ny = y+dy[d];

        if(CanGo(nx, ny)) {
            visited[nx][ny] = true;
            cnt++;
            DFS(nx, ny);
        }
    }
}

int main() {
    cin >> n;

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            cin >> grid[i][j];
        }
    }

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            if(grid[i][j] && !visited[i][j]) {
                visited[i][j] = true;
                cnt = 1;
                DFS(i, j);

                people.push_back(cnt);
            }
        }
    }

    cout << (int)people.size() << '\n';
    sort(people.begin(), people.end());
    for(const auto& p : people) {
        cout << p << '\n';
    }
    return 0;
}

File: 250703/마을 구분하기/seperate-village.py
n = int(input())
grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

visited = [[False]*n for _ in range(n)]
villages = []
people = 1

def can_go(x, y):
    return 0<=x<n and 0<=y<n and not visited[x][y] and grid[x][y]

def dfs(x, y):
    global people

    dxs, dys = [-1, 0, 1, 0], [0, 1, 0, -1]

    for dx, dy in zip(dxs, dys):
        nx, ny = x+dx, y+dy

        if can_go(nx, ny):
            visited[nx][ny] = True
            people += 1
            dfs(nx, ny)

for i in range(n):
    for j in range(n):
        if not visited[i][j] and grid[i][j]:
            people = 1
            visited[i][j] = True

            dfs(i, j)

            villages.append(people)

villages.sort()
print(len(villages))
for p in villages:
    print(p)